File: landrop-main/receiver/android/app/src/main/java/com/landrop/receiver/protocol/Delivery.kt
package com.landrop.receiver.protocol

// Delivery Engine — deterministic delivery-request, transfer-flow, integrity-
// verification, and installation-handoff orchestration (protocol spec §8-§26,
// project spec §12-§19, architecture §18, §21). Pure decision logic only:
// validates input, checks the frozen Phase 03 Delivery State Machine, computes
// the event + next state, and emits a side-effect instruction. No I/O, no
// storage, no hashing, and no installation work (protocol spec §20-§23 gate
// VERIFYING -> INSTALLATION_HANDOFF; the digest is computed by the platform
// seam and handed in, and the installer is invoked by the application layer).
//
// Mirrors sender/desktop/src/protocol/delivery.ts and the Rust mirror
// sender/desktop/src-tauri/src/protocol/delivery.rs.

// Local decision-layer diagnostics, never wire ErrorCodes.
enum class DeliveryProblem {
    INVALID_STATE,
    INVALID_TRANSITION,
    NO_REQUEST,
    NO_TRANSFER,
    REQUEST_MISMATCH,
    RESPONSE_MISMATCH,
    TRANSFER_ID_MISMATCH,
    INVALID_TRANSFER_ID,
    INVALID_BYTES,
    SIZE_EXCEEDED,
    SIZE_MISMATCH,
    SIZE_VALIDATION_FAILED,
    INVALID_DIGEST,
    NO_DIGEST,
    INTEGRITY_MISMATCH,
}

// Side-effect instruction executed by the application layer; this module never
// performs I/O.
sealed interface DeliveryAction {
    data object None : DeliveryAction
    data class SendMessage(val message: ControlMessage) : DeliveryAction
    data object BeginStream : DeliveryAction
    data object StopStream : DeliveryAction
    data object InvokeInstaller : DeliveryAction
    data object Cleanup : DeliveryAction
}

data class DeliveryOutcome(
    val ok: Boolean,
    val event: DeliveryEvent?,
    val to: DeliveryState?,
    val action: DeliveryAction,
    val problem: DeliveryProblem?,
)

fun outcomeProblem(problem: DeliveryProblem): DeliveryOutcome =
    DeliveryOutcome(ok = false, event = null, to = null, action = DeliveryAction.None, problem = problem)

fun decisionEvent(decision: Decision): DeliveryEvent = when (decision) {
    Decision.ACCEPT -> DeliveryEvent.ACCEPTED
    Decision.REJECT -> DeliveryEvent.REJECTED
}

fun createDeliveryResponse(requestId: RequestId, decision: Decision): DeliveryResponse =
    DeliveryResponse(requestId, decision)

/// Clamped percentage (0..100). Progress is observation, not a Delivery State
/// (protocol spec §31): the machine stays in TRANSFERRING.
fun calculateTransferProgress(bytesTransferred: Long, totalBytes: Long): Long {
    if (totalBytes <= 0) return 0
    val pct = bytesTransferred.coerceAtLeast(0) * 100 / totalBytes
    return pct.coerceIn(0, 100)
}

// Defense-in-depth parity with the TypeScript engine, which validates the raw
// string at beginTransfer. The TransferId value class already enforces this.
private val transferIdPattern = Regex("[A-Za-z0-9][A-Za-z0-9_-]{0,127}")

class DeliveryEngine(
    private val sessionId: SessionId,
    initialState: DeliveryState = DeliveryState.SESSION_ESTABLISHED,
) {
    var state: DeliveryState = initialState
        private set
    var request: DeliveryRequest? = null
        private set
    var transferId: TransferId? = null
        private set

    private var bytesTransferred: Long = 0

    val declaredSize: Long get() = request?.application?.identity?.sizeBytes ?: 0
    val expectedDigest: String?
        get() = request?.application?.identity?.sha256?.value
    val bytesTransferredCount: Long get() = bytesTransferred
    val progressPercent: Long get() = calculateTransferProgress(bytesTransferred, declaredSize)

    private fun requireNoTerminal(): DeliveryProblem? =
        if (DeliveryStateMachine.isDeliveryTerminal(state)) DeliveryProblem.INVALID_STATE else null

    private fun apply(event: DeliveryEvent, requestedState: DeliveryState): DeliveryOutcome {
        if (state != requestedState) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        return when (val result = DeliveryStateMachine.transition(state, event)) {
            is DeliveryTransition.Accepted -> {
                state = result.to
                DeliveryOutcome(
                    ok = true,
                    event = event,
                    to = result.to,
                    action = DeliveryAction.None,
                    problem = null,
                )
            }
            is DeliveryTransition.Rejected -> outcomeProblem(DeliveryProblem.INVALID_TRANSITION)
        }
    }

    private fun finish(inner: DeliveryOutcome, action: DeliveryAction): DeliveryOutcome =
        inner.copy(action = action)

    /** Sender: register and send a delivery request (spec §8-§13). */
    fun registerDeliveryRequest(request: DeliveryRequest): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.SESSION_ESTABLISHED) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        if (request.sessionId != sessionId) return outcomeProblem(DeliveryProblem.REQUEST_MISMATCH)
        this.request = request
        bytesTransferred = 0
        val applied = apply(DeliveryEvent.REQUEST_SENT, DeliveryState.SESSION_ESTABLISHED)
        return finish(applied, DeliveryAction.SendMessage(request))
    }

    /** Sender: request is on the wire; wait for the Receiver decision. */
    fun awaitDecision(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        return apply(DeliveryEvent.AWAITING_DECISION, DeliveryState.REQUEST_SENT)
    }

    /** Receiver: validate an incoming delivery request (spec §35-§37). */
    fun receiveIncomingRequest(request: DeliveryRequest): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.SESSION_ESTABLISHED) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        if (request.sessionId != sessionId) return outcomeProblem(DeliveryProblem.REQUEST_MISMATCH)
        this.request = request
        bytesTransferred = 0
        // Mirror of the Sender's machine advance: request present -> awaiting user.
        val applied = apply(DeliveryEvent.REQUEST_SENT, DeliveryState.SESSION_ESTABLISHED)
        return finish(applied, DeliveryAction.None)
    }

    /** Receiver: send the decision (spec §10, §14-§15). */
    fun sendDecision(decision: Decision): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.WAITING_FOR_DECISION) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        val activeRequest = request ?: return outcomeProblem(DeliveryProblem.NO_REQUEST)
        val applied = apply(decisionEvent(decision), DeliveryState.WAITING_FOR_DECISION)
        // Spec §14-§15: the Receiver sends delivery_response for BOTH decisions.
        return finish(
            applied,
            DeliveryAction.SendMessage(createDeliveryResponse(activeRequest.requestId, decision)),
        )
    }

    /** Sender: process the Receiver's response. */
    fun processResponse(response: DeliveryResponse): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.WAITING_FOR_DECISION) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        val activeRequest = request ?: return outcomeProblem(DeliveryProblem.NO_REQUEST)
        if (response.requestId != activeRequest.requestId) return outcomeProblem(DeliveryProblem.RESPONSE_MISMATCH)
        return apply(decisionEvent(response.decision), DeliveryState.WAITING_FOR_DECISION)
    }

    /** Expiration (spec §25; the numeric timeout is a recorded gap). */
    fun expire(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        // A rejected application must not emit a stale cleanup instruction.
        val applied = apply(DeliveryEvent.EXPIRED, DeliveryState.WAITING_FOR_DECISION)
        if (!applied.ok) return applied
        return finish(applied, DeliveryAction.Cleanup)
    }

    /** Transfer preparation (spec §16). */
    fun prepareTransfer(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.ACCEPTED) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        val applied = apply(DeliveryEvent.TRANSFER_PREPARED, DeliveryState.ACCEPTED)
        return finish(applied, DeliveryAction.None)
    }

    /** Transfer start (spec §16-§17). */
    fun beginTransfer(id: TransferId): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.TRANSFER_PREPARING) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        if (request == null) return outcomeProblem(DeliveryProblem.NO_REQUEST)
        if (!transferIdPattern.matches(id.value)) return outcomeProblem(DeliveryProblem.INVALID_TRANSFER_ID)
        if (declaredSize <= 0) return outcomeProblem(DeliveryProblem.SIZE_VALIDATION_FAILED)
        transferId = id
        bytesTransferred = 0
        val applied = apply(DeliveryEvent.TRANSFER_STARTED, DeliveryState.TRANSFER_PREPARING)
        return finish(applied, DeliveryAction.BeginStream)
    }

    /** Byte accounting (spec §18 receiver-side size enforcement). */
    fun recordBytes(count: Long): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.TRANSFERRING) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        if (count <= 0) return outcomeProblem(DeliveryProblem.INVALID_BYTES)
        val next = bytesTransferred + count
        if (next > declaredSize) return outcomeProblem(DeliveryProblem.SIZE_EXCEEDED)
        bytesTransferred = next
        return DeliveryOutcome(ok = true, event = null, to = null, action = DeliveryAction.None, problem = null)
    }

    /** Transfer completion (spec §19-§20): stops at VERIFICATION_INITIATED. */
    fun completeTransfer(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.TRANSFERRING) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        if (transferId == null) return outcomeProblem(DeliveryProblem.NO_TRANSFER)
        if (bytesTransferred != declaredSize) return outcomeProblem(DeliveryProblem.SIZE_MISMATCH)
        val applied = apply(DeliveryEvent.VERIFICATION_INITIATED, DeliveryState.TRANSFERRING)
        return finish(applied, DeliveryAction.None)
    }

    /** Receiver: integrity verification (spec §20-§21). The application layer
     *  computes SHA-256 over the received artifact via the platform seam and
     *  hands the lowercase-hex digest in. The engine classifies it against the
     *  declared digest and routes through the frozen machine. */
    fun completeVerification(actualDigest: String): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.VERIFYING) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        val request = request ?: return outcomeProblem(DeliveryProblem.NO_REQUEST)
        val expected = request.application.identity.sha256.value
        // Defensive (unreachable: Sha256 validates 64 hex chars at construction).
        if (expected.isEmpty()) return outcomeProblem(DeliveryProblem.NO_DIGEST)
        if (!Integrity.isValidSha256(actualDigest)) return outcomeProblem(DeliveryProblem.INVALID_DIGEST)
        return if (Integrity.hashesMatch(expected, actualDigest)) {
            val applied = apply(DeliveryEvent.VERIFIED, DeliveryState.VERIFYING)
            finish(applied, DeliveryAction.None)
        } else {
            // Protocol §21: mismatch -> FAILED, installation blocked, cleanup.
            val applied = apply(DeliveryEvent.INTEGRITY_MISMATCH, DeliveryState.VERIFYING)
            finish(applied, DeliveryAction.Cleanup).let {
                it.copy(problem = DeliveryProblem.INTEGRITY_MISMATCH)
            }
        }
    }

    /** Receiver: installation handoff (spec §23, project spec §26). */
    fun prepareInstall(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.VERIFIED) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        val applied = apply(DeliveryEvent.INSTALL_PREPARED, DeliveryState.VERIFIED)
        return finish(applied, DeliveryAction.None)
    }

    /** The application layer calls this immediately before invoking the
     *  platform installer; the emitted instruction tells it exactly that. */
    fun handoffInstall(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.INSTALL_READY) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        val applied = apply(DeliveryEvent.HANDOFF_INITIATED, DeliveryState.INSTALL_READY)
        return finish(applied, DeliveryAction.InvokeInstaller)
    }

    /** Called only when the platform's supported package-install mechanism has
     *  genuinely reported completion. Nothing fabricates this result. */
    fun installationCompleted(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.INSTALLATION_HANDOFF) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        // Terminal state -> cleanup instruction (artifact/session lifecycle).
        val applied = apply(DeliveryEvent.COMPLETED, DeliveryState.INSTALLATION_HANDOFF)
        return finish(applied, DeliveryAction.Cleanup)
    }

    /** Platform reports the package installer is unavailable, before or during
     *  handoff (project spec §24 INSTALLATION_UNAVAILABLE). */
    fun installationUnavailable(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.INSTALL_READY && state != DeliveryState.INSTALLATION_HANDOFF) {
            return outcomeProblem(DeliveryProblem.INVALID_STATE)
        }
        val applied = apply(DeliveryEvent.INSTALLATION_UNAVAILABLE, state)
        if (!applied.ok) return applied
        return finish(applied, DeliveryAction.Cleanup)
    }

    /** Cancellation (spec §24; machine-gated). */
    fun cancel(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        val applied = apply(DeliveryEvent.CANCELLED, state)
        if (!applied.ok) return applied
        return finish(applied, DeliveryAction.Cleanup)
    }

    /** Incoming transfer_cancel from the Sender (spec §24). */
    fun processTransferCancel(id: TransferId): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        val active = transferId ?: return outcomeProblem(DeliveryProblem.NO_TRANSFER)
        if (active != id) return outcomeProblem(DeliveryProblem.TRANSFER_ID_MISMATCH)
        val applied = apply(DeliveryEvent.CANCELLED, state)
        if (!applied.ok) return applied
        return finish(applied, DeliveryAction.StopStream)
    }

    /** Transfer inactivity timeout (spec §26). */
    fun transferTimeout(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        if (state != DeliveryState.TRANSFERRING) return outcomeProblem(DeliveryProblem.INVALID_STATE)
        val applied = apply(DeliveryEvent.TRANSFER_TIMED_OUT, DeliveryState.TRANSFERRING)
        if (!applied.ok) return applied
        return finish(applied, DeliveryAction.Cleanup)
    }

    /** Connection loss at any pre-terminal stage (spec §27). */
    fun connectionLost(): DeliveryOutcome {
        requireNoTerminal()?.let { return outcomeProblem(it) }
        val applied = apply(DeliveryEvent.CONNECTION_LOST, state)
        if (!applied.ok) return applied
        return finish(applied, DeliveryAction.Cleanup)
    }
}

File: landrop-main/sender/desktop/src-tauri/src/protocol/delivery.rs
//! Delivery Engine — deterministic delivery-request, transfer-flow, integrity-
//! verification, and installation-handoff orchestration (protocol spec §8-§26,
//! project spec §12-§19, architecture §18, §21). Pure decision logic only:
//! validates input, checks the frozen Phase 03 Delivery State Machine, computes
//! the event + next state, and emits a side-effect instruction. No I/O, no
//! storage, no hashing, and no installation work (protocol §20-§23 gate
//! VERIFYING -> INSTALLATION_HANDOFF; the digest is computed by the platform
//! seam and handed in, and the installer is invoked by the application layer).
//!
//! Mirrors sender/desktop/src/protocol/delivery.ts and the Kotlin mirror
//! receiver/android/.../protocol/Delivery.kt.

use super::{
    hashes_match, is_delivery_terminal, is_valid_sha256, transition, ControlMessage, Decision,
    DeliveryEvent, DeliveryRequest, DeliveryResponse, DeliveryState, DeliveryTransition,
    RequestId, SessionId, TransferId,
};

// Local decision-layer diagnostics, never wire ErrorCodes (same guarantee as
// the TypeScript mirror: no new wire error codes are invented).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DeliveryProblem {
    InvalidState,
    InvalidTransition,
    NoRequest,
    NoTransfer,
    RequestMismatch,
    ResponseMismatch,
    TransferIdMismatch,
    InvalidTransferId,
    InvalidBytes,
    SizeExceeded,
    SizeMismatch,
    SizeValidationFailed,
    InvalidDigest,
    NoDigest,
    IntegrityMismatch,
}

// Side-effect instruction executed by the application layer; this module never
// performs I/O.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DeliveryAction {
    None,
    SendMessage(ControlMessage),
    BeginStream,
    StopStream,
    InvokeInstaller,
    Cleanup,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeliveryOutcome {
    pub ok: bool,
    pub event: Option<DeliveryEvent>,
    pub to: Option<DeliveryState>,
    pub action: DeliveryAction,
    pub problem: Option<DeliveryProblem>,
}

pub fn outcome_problem(problem: DeliveryProblem) -> DeliveryOutcome {
    DeliveryOutcome {
        ok: false,
        event: None,
        to: None,
        action: DeliveryAction::None,
        problem: Some(problem),
    }
}

pub fn decision_event(decision: Decision) -> DeliveryEvent {
    match decision {
        Decision::Accept => DeliveryEvent::Accepted,
        Decision::Reject => DeliveryEvent::Rejected,
    }
}

pub fn create_delivery_response(request_id: &RequestId, decision: Decision) -> DeliveryResponse {
    DeliveryResponse {
        request_id: request_id.clone(),
        decision,
    }
}

/// Clamped percentage (0..=100). Progress is observation, not a Delivery State
/// (protocol spec §31): the machine stays in Transferring.
pub fn calculate_transfer_progress(bytes_transferred: i64, total_bytes: i64) -> i64 {
    if total_bytes <= 0 {
        return 0;
    }
    (bytes_transferred.max(0) * 100 / total_bytes).clamp(0, 100)
}

// Defense-in-depth parity with the TypeScript engine, which validates the raw
// string at beginTransfer. The TransferId newtype already enforces this grammar
// at construction, so this is always true for values that reach the engine.
fn transfer_id_ok(id: &TransferId) -> bool {
    let value = id.as_str();
    (1..=128).contains(&value.len())
        && value.as_bytes()[0].is_ascii_alphanumeric()
        && value.bytes().all(|c| c.is_ascii_alphanumeric() || c == b'_' || c == b'-')
}

pub struct DeliveryEngine {
    session_id: SessionId,
    state: DeliveryState,
    request: Option<DeliveryRequest>,
    transfer_id: Option<TransferId>,
    bytes_transferred: i64,
}

impl DeliveryEngine {
    pub fn new(session_id: SessionId) -> Self {
        Self::with_state(session_id, DeliveryState::SessionEstablished)
    }

    pub fn with_state(session_id: SessionId, state: DeliveryState) -> Self {
        Self {
            session_id,
            state,
            request: None,
            transfer_id: None,
            bytes_transferred: 0,
        }
    }

    pub fn state(&self) -> DeliveryState {
        self.state
    }

    pub fn request(&self) -> Option<&DeliveryRequest> {
        self.request.as_ref()
    }

    pub fn transfer_id(&self) -> Option<&TransferId> {
        self.transfer_id.as_ref()
    }

    pub fn declared_size(&self) -> i64 {
        self.request
            .as_ref()
            .map(|r| r.application.identity.size_bytes)
            .unwrap_or(0)
    }

    /// Expected digest the Sender declared in the delivery request (project
    /// spec §14). The receiver verifies its independently computed digest
    /// against it.
    pub fn expected_digest(&self) -> Option<&str> {
        self.request
            .as_ref()
            .map(|r| r.application.identity.sha256.as_str())
    }

    pub fn bytes_transferred(&self) -> i64 {
        self.bytes_transferred
    }

    pub fn progress_percent(&self) -> i64 {
        calculate_transfer_progress(self.bytes_transferred, self.declared_size())
    }

    fn require_no_terminal(&self) -> Option<DeliveryProblem> {
        if is_delivery_terminal(self.state) {
            Some(DeliveryProblem::InvalidState)
        } else {
            None
        }
    }

    fn apply(&mut self, event: DeliveryEvent, requested_state: DeliveryState) -> DeliveryOutcome {
        if self.state != requested_state {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        match transition(self.state, event) {
            DeliveryTransition::Accepted(to) => {
                self.state = to;
                DeliveryOutcome {
                    ok: true,
                    event: Some(event),
                    to: Some(to),
                    action: DeliveryAction::None,
                    problem: None,
                }
            }
            DeliveryTransition::Rejected { .. } => outcome_problem(DeliveryProblem::InvalidTransition),
        }
    }

    fn finish(inner: DeliveryOutcome, action: DeliveryAction) -> DeliveryOutcome {
        DeliveryOutcome { action, ..inner }
    }

    /// Sender: register and send a delivery request (spec §8-§13).
    pub fn register_delivery_request(&mut self, request: DeliveryRequest) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::SessionEstablished {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        if request.session_id != self.session_id {
            return outcome_problem(DeliveryProblem::RequestMismatch);
        }
        self.request = Some(request.clone());
        self.bytes_transferred = 0;
        let applied = self.apply(DeliveryEvent::RequestSent, DeliveryState::SessionEstablished);
        Self::finish(
            applied,
            DeliveryAction::SendMessage(ControlMessage::DeliveryRequest(request)),
        )
    }

    /// Sender: request is on the wire; wait for the Receiver decision.
    pub fn await_decision(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        self.apply(DeliveryEvent::AwaitingDecision, DeliveryState::RequestSent)
    }

    /// Receiver: validate an incoming delivery request (spec §35-§37).
    pub fn receive_incoming_request(&mut self, request: DeliveryRequest) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::SessionEstablished {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        if request.session_id != self.session_id {
            return outcome_problem(DeliveryProblem::RequestMismatch);
        }
        self.request = Some(request);
        self.bytes_transferred = 0;
        // Mirror of the Sender's machine advance: request present -> awaiting user.
        let applied = self.apply(DeliveryEvent::RequestSent, DeliveryState::SessionEstablished);
        Self::finish(applied, DeliveryAction::None)
    }

    /// Receiver: send the decision (spec §10, §14-§15).
    pub fn send_decision(&mut self, decision: Decision) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::WaitingForDecision {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let Some(request) = &self.request else {
            return outcome_problem(DeliveryProblem::NoRequest);
        };
        let request_id = request.request_id.clone();
        let applied = self.apply(decision_event(decision), DeliveryState::WaitingForDecision);
        // Spec §14-§15: the Receiver sends delivery_response for BOTH decisions.
        Self::finish(
            applied,
            DeliveryAction::SendMessage(ControlMessage::DeliveryResponse(
                create_delivery_response(&request_id, decision),
            )),
        )
    }

    /// Sender: process the Receiver's response.
    pub fn process_response(&mut self, response: DeliveryResponse) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::WaitingForDecision {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let Some(request) = &self.request else {
            return outcome_problem(DeliveryProblem::NoRequest);
        };
        if response.request_id != request.request_id {
            return outcome_problem(DeliveryProblem::ResponseMismatch);
        }
        self.apply(decision_event(response.decision), DeliveryState::WaitingForDecision)
    }

    /// Expiration (spec §25; the numeric timeout is a recorded gap).
    pub fn expire(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        // A rejected application must not emit a stale cleanup instruction.
        let applied = self.apply(DeliveryEvent::Expired, DeliveryState::WaitingForDecision);
        if !applied.ok {
            return applied;
        }
        Self::finish(applied, DeliveryAction::Cleanup)
    }

    /// Transfer preparation (spec §16).
    pub fn prepare_transfer(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::Accepted {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let applied = self.apply(DeliveryEvent::TransferPrepared, DeliveryState::Accepted);
        Self::finish(applied, DeliveryAction::None)
    }

    /// Transfer start (spec §16-§17).
    pub fn begin_transfer(&mut self, transfer_id: TransferId) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::TransferPreparing {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        if self.request.is_none() {
            return outcome_problem(DeliveryProblem::NoRequest);
        }
        if !transfer_id_ok(&transfer_id) {
            return outcome_problem(DeliveryProblem::InvalidTransferId);
        }
        let declared = self.declared_size();
        if declared <= 0 {
            return outcome_problem(DeliveryProblem::SizeValidationFailed);
        }
        self.transfer_id = Some(transfer_id);
        self.bytes_transferred = 0;
        let applied = self.apply(DeliveryEvent::TransferStarted, DeliveryState::TransferPreparing);
        Self::finish(applied, DeliveryAction::BeginStream)
    }

    /// Byte accounting (spec §18 receiver-side size enforcement).
    pub fn record_bytes(&mut self, count: i64) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::Transferring {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        if count <= 0 {
            return outcome_problem(DeliveryProblem::InvalidBytes);
        }
        let next = self.bytes_transferred.saturating_add(count);
        if next > self.declared_size() {
            return outcome_problem(DeliveryProblem::SizeExceeded);
        }
        self.bytes_transferred = next;
        DeliveryOutcome {
            ok: true,
            event: None,
            to: None,
            action: DeliveryAction::None,
            problem: None,
        }
    }

    /// Transfer completion (spec §19-§20): stops at VerificationInitiated.
    pub fn complete_transfer(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::Transferring {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        if self.transfer_id.is_none() {
            return outcome_problem(DeliveryProblem::NoTransfer);
        }
        if self.bytes_transferred != self.declared_size() {
            return outcome_problem(DeliveryProblem::SizeMismatch);
        }
        let applied = self.apply(DeliveryEvent::VerificationInitiated, DeliveryState::Transferring);
        Self::finish(applied, DeliveryAction::None)
    }

    /// Receiver: integrity verification (spec §20-§21). The application layer
    /// computes SHA-256 over the received artifact via the platform seam and
    /// hands the lowercase-hex digest in. The engine classifies it against the
    /// declared digest and routes through the frozen machine.
    pub fn complete_verification(&mut self, actual_digest: &str) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::Verifying {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let Some(request) = &self.request else {
            return outcome_problem(DeliveryProblem::NoRequest);
        };
        let expected = &request.application.identity.sha256;
        // Unreachable in practice: the Sha256 newtype guarantees 64 hex chars,
        // so an empty/absent declared digest cannot exist. Kept for defensive
        // parity with the TS and Kotlin mirrors, whose boundaries accept the
        // raw string.
        if expected.as_str().is_empty() {
            return outcome_problem(DeliveryProblem::NoDigest);
        }
        if !is_valid_sha256(actual_digest) {
            return outcome_problem(DeliveryProblem::InvalidDigest);
        }
        if hashes_match(expected.as_str(), actual_digest) {
            let applied = self.apply(DeliveryEvent::Verified, DeliveryState::Verifying);
            Self::finish(applied, DeliveryAction::None)
        } else {
            // Protocol §21: mismatch -> FAILED, installation blocked, cleanup.
            let applied = self.apply(DeliveryEvent::IntegrityMismatch, DeliveryState::Verifying);
            let outcome = Self::finish(applied, DeliveryAction::Cleanup);
            DeliveryOutcome {
                problem: Some(DeliveryProblem::IntegrityMismatch),
                ..outcome
            }
        }
    }

    /// Receiver: installation handoff (spec §23, project spec §26).
    pub fn prepare_install(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::Verified {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let applied = self.apply(DeliveryEvent::InstallPrepared, DeliveryState::Verified);
        Self::finish(applied, DeliveryAction::None)
    }

    /// The application layer calls this immediately before invoking the
    /// platform installer; the emitted instruction tells it exactly that.
    pub fn handoff_install(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::InstallReady {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let applied = self.apply(DeliveryEvent::HandoffInitiated, DeliveryState::InstallReady);
        Self::finish(applied, DeliveryAction::InvokeInstaller)
    }

    /// Called only when the platform's supported package-install mechanism has
    /// genuinely reported completion. Nothing fabricates this result.
    pub fn installation_completed(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::InstallationHandoff {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        // Terminal state -> cleanup instruction (artifact/session lifecycle).
        let applied = self.apply(DeliveryEvent::Completed, DeliveryState::InstallationHandoff);
        Self::finish(applied, DeliveryAction::Cleanup)
    }

    /// Platform reports the package installer is unavailable, before or during
    /// handoff (project spec §24 INSTALLATION_UNAVAILABLE).
    pub fn installation_unavailable(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::InstallReady
            && self.state != DeliveryState::InstallationHandoff
        {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let applied = self.apply(DeliveryEvent::InstallationUnavailable, self.state);
        if !applied.ok {
            return applied;
        }
        Self::finish(applied, DeliveryAction::Cleanup)
    }

    /// Cancellation (spec §24; machine-gated).
    pub fn cancel(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        let applied = self.apply(DeliveryEvent::Cancelled, self.state);
        if !applied.ok {
            return applied;
        }
        Self::finish(applied, DeliveryAction::Cleanup)
    }

    /// Incoming transfer_cancel from the Sender (spec §24).
    pub fn process_transfer_cancel(&mut self, transfer_id: TransferId) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        let Some(active) = &self.transfer_id else {
            return outcome_problem(DeliveryProblem::NoTransfer);
        };
        if *active != transfer_id {
            return outcome_problem(DeliveryProblem::TransferIdMismatch);
        }
        let applied = self.apply(DeliveryEvent::Cancelled, self.state);
        if !applied.ok {
            return applied;
        }
        Self::finish(applied, DeliveryAction::StopStream)
    }

    /// Transfer inactivity timeout (spec §26).
    pub fn transfer_timeout(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        if self.state != DeliveryState::Transferring {
            return outcome_problem(DeliveryProblem::InvalidState);
        }
        let applied = self.apply(DeliveryEvent::TransferTimedOut, DeliveryState::Transferring);
        if !applied.ok {
            return applied;
        }
        Self::finish(applied, DeliveryAction::Cleanup)
    }

    /// Connection loss at any pre-terminal stage (spec §27).
    pub fn connection_lost(&mut self) -> DeliveryOutcome {
        if let Some(terminal) = self.require_no_terminal() {
            return outcome_problem(terminal);
        }
        let applied = self.apply(DeliveryEvent::ConnectionLost, self.state);
        if !applied.ok {
            return applied;
        }
        Self::finish(applied, DeliveryAction::Cleanup)
    }
}